### PR TITLE
Small performance improvement in Tracking

### PR DESCRIPTION
**Pull request by VinInn**
This PR addresses some performance hotspots in tracking code.
the main improvement is to reduce the size of the "cluster" for the Template computation.
Negligible regressions are expected due to numerics and to a different behavior in PixelCPETemplateReco for some extreme cases (large cluster)

same as #13382 w/o MagneticField

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by slava77** [comment]
Looking at the jenkins comparisons, there are changes in quite a few plots.
Please provide the tracking MTV plots (we usually look at 25202, but if there is more, it would be nice to see as well).

**Comment by VinInn** [comment]
I've in hand only those with mag-field mods
and in CMSSW_8_1_X_2016-02-22-2300 with my standard config I get
cannot import name iterativeCone5GenJetsNoNu
any clue?

**Comment by slava77** [comment]
https://cmssdt.cern.ch/SDT/html/showIB.html
CMSSW_8_1_X_2016-02-22-2300 had the same errors in the slc6gcc493 build.
It was probably an afs glitch at install step or something like that.

**Comment by slava77** [comment]
Looking at 134.911 in jenkins
wf134 911_cluprob_bpix1_box
The same logscale:
wf134 911_cluprob_bpix1_log_sign673vsorig

Now about a half of clusters have a probability ~1
and most of them moved there from the range of moderate probs (>~0.1).
Is this roughly the expected? I thought that the size reduction for CPE would affect the tails more.

**Comment by VinInn** [comment]
1) So I have to wait for a new IB and rebuild my test area.
2) for what concern cluster probability: there was (is) a loop over the whole matrix to check for bad pixels,
It looks suspicious to me that it had to loop "outside the actual cluster". Now it does not anymore (by construction). 
Maybe we have to ask Morris
p.s. cluster probability is not used. anywhere.

**Comment by slava77** [comment]
@VinInn since this changes performance of local reco, some confirmation from the DPG is needed.
@boudoul @dkotlins @veszpv 
Maybe this can be shown/discussed in the joint  POG/DPG meeting tomorrow?

**Comment by dkotlins** [comment]
As Vincenzo said, we do not use these probabilities very much.
I will forward the plots to Morris so he can comment.
Danek

On 25 Feb 2016, at 19:50, Slava Krutelyov <notifications@github.com<mailto:notifications@github.com>> wrote:


@VinInn<https://github.com/VinInn> since this changes performance of local reco, some confirmation from the DPG is needed.
@boudoul<https://github.com/boudoul> @dkotlins<https://github.com/dkotlins> @veszpv<https://github.com/veszpv>
Maybe this can be shown/discussed in the joint POG/DPG meeting tomorrow?

**Comment by slava77** [comment]
Some "random" plots:

1313 (3 TeV dijet)
minor increase in PV errors
wf1313_sign673-pass-abb287avsorig_pverrx
errors increase also in b-tagging variables

minor decrease in track chi2
all_sign673-pass-abb287avsorig_qcd13tevpt3ts3t5wf1313p0c_minrecotracks_generaltracks__reco_obj_normalizedchi2 29
(most jets in this sample are ~central ==> low stats at higher eta)
wf1313_sign673-pass-abb287avsorig_gentk_chi2_eta
this turns into less flat track prob
wf1313_sign673-pass-abb287avsorig_gentk_prob

Some moderate decrease in track charge mis-id
wf1313_sign673-pass-abb287avsorig_gentk_qmisid

From the HN it looks like this PR will be closed or redone.
So, I'm not continuing with checks much further.

File: RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco.cc [hunk 1]
```diff
@@ -196,44 +191,25 @@ int SiPixelTemplateReco::PixelTempReco2D(int id, float cotalpha, float cotbeta,
...
+	auto const xdouble = cluster.xdouble;
+	auto const ydouble = cluster.ydouble;
 
-	for(j=0; j<nclusx; ++j)
-    for(i=0; i<nclusy; ++i)
-		  if(cluster[j][i] > 0) {cluster[j][i] *= qscale;}
+// First, rescale all pixel charges and compute total charge
+        qtotal = 0.f;
+	for(i=0; i<nclusx*nclusy; ++i)
+	   cluster.matrix[i] *= qscale; qtotal +=cluster.matrix[i];
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco.cc]: missing brace!

**Comment by VinInn** [comment]
and needs rebase!

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by VinInn** [comment]
mtv pu35 ttbar
http://innocent.home.cern.ch/innocent/RelVal/pu35_TkImprove810/

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/13448	025a5ae
- changes in the code are in line with the PR description and some clarifications in the follow up discussion
- jenkins tests pass and comparisons with baseline show small differences consistent with numerical precision level changes
- MTV plots provided by @VinInn  show negligible changes in the general tracks
- tested locally  on larger samples in CMSSW_8_1_X_2016-02-22-1100 (some cherry-picking involved for historical reasons)
    - general tracks are minimally affected
    - most noticeable (but apparently not significant) changes are in allConversions and in gsfElectronTracks at per mille level on yields and more frequent track value changes
    - the effects on dowsntream objects are small
    - CPU use shows a somewhat noticeable decrease: \*Track\* modules in reco now take about 1.5% less time; the total reco step is about 0.7% faster.

**Comment by civanch** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**